library/src/main/scala/sbt/SendEventService.scala

```diff
@@ -7,6 +7,20 @@ sealed trait SendEventService {
   def sendEvent[T: Pickler](event: T): Unit
 }
 
+object SendEventService {
+  private val detachedKey = AttributeKey[SendEventService]("SendEventService that sends DetachedEvent")
+
+  /**
+   * To send a detached plugin event, do getDetached(state).foreach(_.sendEvent(event)).
+   * This should never be done from inside a task or background job where a better
+   * SendEventService is available already. "Detached" means not associated with a task
+   * or job.
+   */
+  def getDetached(state: State): Option[SendEventService] = state get detachedKey
```

Comment:
extract looks like a totally different deal that wants to load the session and the build structure - Project.getOrError is more what's wanted here, but it is just `state get key getOrElse sys.error`